### PR TITLE
Migrate reminders endpoint from the site API, introduce model improvements

### LINKED ISSUE #22: Migrate `reminders` endpoint from the site API
At this point, I believe it's time we started migrating a few of the endpoints, there is an issue regarding the `off-topic names` endpoint, however I think we can also migrate the `reminders` endpoint, to see a more complicated one in this new environment as well.
(Although, the Alembic migration environment hasn't been set up just yet, a PR should come up anytime now.)
I would be glad If I could implement( or more like migrate) this one myself. (If it's possible, of course)

### PR DESCRIPTION
**Pull request by D0rs4n**
Closes #22
- Add reminder endpoint package to the project:
 - Add reminder dependencies
 - Add reminder schemas
 - Add reminder endpoints
- Add reminder tests to the project.
 - The tests use mocked DB session calls, to avoid using a test DB.
- Introduce changes in the flake8 config, and main to be comaptible with the test suites.
- Introduce minor changes in the Database models. (Setting defaults)

### CODE REVIEW

File: .github/workflows/lint-test.yml [hunk 1]
```diff
@@ -24,7 +24,7 @@ jobs:
       PRE_COMMIT_HOME: ${{ github.workspace }}/.cache/pre-commit-cache
 
       AUTH_TOKEN: ci-token
-      DATABASE_URL: postgres://pysite:pysite@postgres:5432/pysite
+      TESTING: true
```
Comment (Shivansh-007) [inline .github/workflows/lint-test.yml:27]: We shouldn't need this or any of the test patching, we can just use a local database for local testing and the actions, we already have a database service, see:

https://github.com/python-discord/api/blob/6df16aea2868c39a60b5e42189e507be6ebbfc35/.github/workflows/lint-test.yml#L29-L44

File: api/endpoints/reminder/reminder_schemas.py [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+    content: str
+    expiration: datetime  # ISO-formatted datetime
+    id: int
+    channel_id: int
+    jump_url: str
+
+    @validator("expiration")
+    def parse_expiration(cls, value: datetime) -> str:  # noqa N805
+        """A parser that transforms datetimes into isoformat."""
+        return value.isoformat()
```
Comment (Shivansh-007) [inline api/endpoints/reminder/reminder_schemas.py:22]: We should raise a value error if this fails to be formatted.

File: tests/endpoints/test_reminders.py [hunk 3]
```diff
@@ -0,0 +1,272 @@
...
+from api.endpoints.dependencies.database import create_database_session
+from api.endpoints.reminder.reminder_schemas import ReminderResponse
+from api.main import app
+
+client = TestClient(app)
+
+AUTH_HEADER = {"Authorization": f"Bearer {settings.auth_token}"}
+
+
+class TestUnauthedReminderAPI:
```
Comment (Shivansh-007) [inline tests/endpoints/test_reminders.py:17]: It would be better if these inherit `unittest.Testcase` and use its methods rather than defining a dead `case = unittest.TestCase()` and using it. Later, you could also remove the `Test` prefix from this.

File: .github/workflows/lint-test.yml (hunk 1 above)
Comment (D0rs4n) [reply .github/workflows/lint-test.yml:27]: It was discussed on discord, that we should try not to use a database for testing so that we can implement further features in the future.

File: tests/endpoints/test_reminders.py (hunk 3 above)
Comment (D0rs4n) [reply tests/endpoints/test_reminders.py:17]: I added this so that I can stick with the general pytest syntax, it integrates pretty well. The 'Test' in the name is a pytest convention.

**Review by jchristgit — CHANGES_REQUESTED** [review]
Using a mocked test database in the tests is a very huge stepback from the existing Django API and renders all of the added tests far less effective. Please re-add the test database and integrate it with our CI properly such that we can validate that the endpoints works with the database as wanted.

**Comment by D0rs4n** [comment]
> Using a mocked test database in the tests is a very huge stepback from the existing Django API and renders all of the added tests far less effective. Please re-add the test database and integrate it with our CI properly such that we can validate that the endpoints works with the database as wanted.

Alright, I was gonna do this initially however @HassanAbouelela [mentioned](https://discord.com/channels/267624335836053506/635950537262759947/904038420547002438) that it would make the possibility of implementing features such as  parallelization impossible, and that I shouldn't use a test database. 
I think it's worth a discussion.

**Comment by HassanAbouelela** [comment]
@jchristgit My reasoning for not adding the test database is that we don't really need it for our tests. I've looked over a few of the current site tests, and for the api ones, it's mostly just "Put data in DB, pull data from DB, call functions with that data." If you notice there, that whole putting into DB, and taking out is completely redundant. In both cases you're creating a predetermined instance of a model, and injecting it into a route. We don't need a DB to do that.

On the other hand, having a DB does make it more difficult to do certain things (such as parallelization as mentioned, but there are others). For instance, a test DB that is shared throughout the entire test suite makes for a more fragile suite, as your tests can have unexpected run conditions from other tests.

TL;DR: Imo, having a DB in our test suites doesn't solve much, but does make testing more complex.

File: api/endpoints/reminder/reminder_schemas.py (hunk 2 above)
Comment (D0rs4n) [reply api/endpoints/reminder/reminder_schemas.py:22]: Before it gets passed to the parser, pydantic will make sure that it is a datetime object, so I'm not sure I see where it could fail.

**Comment by jchristgit** [comment]
> @jchristgit My reasoning for not adding the test database is that we don't really need it for our tests. I've looked over a few of the current site tests, and for the api ones, it's mostly just "Put data in DB, pull data from DB, call functions with that data." If you notice there, that whole putting into DB, and taking out is completely redundant. In both cases you're creating a predetermined instance of a model, and injecting it into a route. We don't need a DB to do that.

I can understand your reasoning. For the majority of tests, it's true that this is redundant. However, by using a mock database, we could miss big problems such as broken inserts due to failed constraints, wrong types (or issues such as character truncation errors), foreign key problems, and more. For simple routes that are just essentially CRUD of routes, we could perhaps create some helper functions to perform the "create, test read, update, test updated, delete, test deleted" sequence of events whilst saving us from setting up mocks.

> On the other hand, having a DB does make it more difficult to do certain things (such as parallelization as mentioned, but there are others). For instance, a test DB that is shared throughout the entire test suite makes for a more fragile suite, as your tests can have unexpected run conditions from other tests.

We can deal with this just fine as we are luckily using a sane database. Allowing parallel execution in our tests would basically just require:

- Not relying on shared test information (such as fixed IDs), but having the database generate them for us
- Using transactions to allow for parallel testing

**Comment by D0rs4n** [comment]
So, I see valid points on both sides. 
What's the consensus? 
I can push a draft solution how the testing using a test DB would look like. (Or, how it would not )

**Comment by D0rs4n** [comment]
Because of recent discussions and changes, I decided to close this pull request, and rewrite the changes from (almost) scratch.
